packages/react-components/react-motions-preview/src/utils/groups/mergeChildMappings.ts

```diff
@@ -0,0 +1,56 @@
+import type { PresenceGroupChildMapping } from './types';
+
+/**
+ * When you're adding or removing children some may be added or removed in the same render pass. We want to show *both*
+ * since we want to simultaneously animate elements in and out. This function takes a previous set of keys and a new set
+ * of keys and merges them with its best guess of the correct ordering.
+ */
+export function mergeChildMappings(
+  prevMapping: PresenceGroupChildMapping,
+  nextMapping: PresenceGroupChildMapping,
+): PresenceGroupChildMapping {
+  function getValueForKey(key: string) {
+    return key in nextMapping ? nextMapping[key] : prevMapping[key];
+  }
+
+  // For each key of `next`, the list of keys to insert before that key in
+  // the combined list
+  const nextKeysPending: Record<string, string[]> = {};
+  let pendingKeys: string[] = [];
+
+  // eslint-disable-next-line guard-for-in
+  for (const prevKey in prevMapping) {
```

Comment:
There's some assumptions made here about the order of children between states, I would add tests to make sure that it's not forgotten